lambda/translcore.ml

```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
       match type_comp with 
       | In (id, _, e2) -> 
-          let in_ = (transl_exp ~scopes e2) in
-          let len = Lprim( (Parraylength(Pgenarray)), [in_], Loc_unknown) in
-          let index = Ident.create_local "index" in    
-          if is_base then
-            let loop = 
-              Lfor(index, (int 0), (l_int_op Psubint len (int 1)), Upto, 
-              construct_base_body (
-                  l_array_set 
-                    (Lvar(res)) 
-                    (Lvar(counter)) 
-                    (Llet(Strict,Pintval, id, 
-                      (l_array_get  in_ (Lvar(index))), 
-                      (transl_exp ~scopes  body)))
-                  ) counter)
-            in 
-            loop, len
-          else 
-            let inner_body, inner_len = nested_for ls res counter in
-            let loop = 
-              Lfor(index, 
-                (int 0), (l_int_op Psubint len (int 1)), Upto, 
-                (Llet(
-                  Strict,Pintval, id, (l_array_get  in_ (Lvar(index))), 
-                  inner_body)))
-            in 
-            loop, l_int_op Pmulint inner_len len
-            
+        let in_ = (transl_exp ~scopes e2) in
+        let len = Lprim( (Parraylength(Pgenarray)), [in_], Loc_unknown) in
+        let index = Ident.create_local "index" in    
+        Lfor(index, (int 0), (transl_int_op Psubint len (int 1)), Upto, 
+            Llet(
+            Strict,Pintval, id, transl_array_get  in_ (Lvar(index)),
+            body)), len
       | From_to(id, _, e2, e3, dir) ->  
         let from = (transl_exp ~scopes e2) in
         let to_ = (transl_exp ~scopes e3) in
         let low, high = 
           match dir with 
           | Upto -> from, to_
           | Downto -> to_, from in
-        let len =  l_int_op Psubint (l_int_op Paddint high (int 1)) low in
-        if is_base then
-          let loop =
-            Lfor(id,from, to_, dir, 
-              Lsequence( 
-                l_array_set (Lvar(res)) (Lvar(counter)) 
-                  (transl_exp ~scopes  body),
-                Lassign(counter, l_int_op Paddint (int 1) (Lvar(counter))))) 
-          in
-          loop, len
+        let len = 
+          transl_int_op Psubint (transl_int_op Paddint high (int 1)) low 
+        in
+        Lfor(id,from, to_, dir, body), len
+    in
+    (*Translates one block into a block of for loops.*)
+    let transl_loops block base_body = 
+      let base_loop = transl_loop ~type_comp:(List.hd block) ~body:base_body in
+      List.fold_left (fun (body, len) type_comp ->
+        let new_body, new_len = transl_loop ~type_comp ~body in
+        new_body, transl_int_op Pmulint new_len len 
+        ) base_loop (List.tl block)
+    in       
+    (*The block created here is the most inner block. This needs to execute the
+      execute the actual body of the comprehension and write it into the 
+      correct result array.*)   
+    let transl_base_block block gloabl_counter is_only_block =  
```

Comment:
I think that the only real difference between `transl_base_block` and `transl_block` is that `transl_base_block` has to run the first iteration before it makes the array since it might be creating a `Pgenarray` (i.e. array that may or may not hold floats) whilst `transl_block` knows it is creating a `Paddrarray` (i.e. an array that is definitely doesn't hold floats).

I think that it would be slightly better to have a single recursive `transl_blocks` function that took an `array_kind` as a parameter, saying what kind of array was being created. If the kind is `Pgenarray` then it does the extra work to run the first iteration before making the array, otherwise it just does the simple version. When this function recursives to handle nested blocks it would pass in `Paddrarray` since it would know that the nested arrays always held pointers to arrays.